Source/WebCore/platform/graphics/FlatDynamicRangeLimit.cpp

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FlatDynamicRangeLimit.h"
+
+#include <wtf/text/TextStream.h>
+
+namespace WebCore {
+
+FlatDynamicRangeLimit FlatDynamicRangeLimit::flatten(float standard, float constrainedHigh, float noLimit)
+{
+    float sum = standard + constrainedHigh + noLimit;
```

Comment:
What are the expected values of `standard`, `constrainedHigh`, `high`?
Can they be negative?
Can `sum` overflow?
Can you please either add `ASSERT` statements if we know the expected values or make an early return if they are wrong?

Reply:
Coming from Style, they should always be between 0 and 1 (or 0 and 100, not important) and add up to 1 (or 100).
I think I'll add an `ASSERT` and an early return.
Then since each value will be in [0,sum] I won't need `safeFPDivision` and `clamp` anymore, but I'll still need to catch `areEssentiallyEqual(sum, 0)`.